design-proposals/klip-42-schema-migrations-tool.md

```diff
@@ -0,0 +1,442 @@
+- V2__Add_users.sql        # a new version to migrate (v2)
+- U2__Add_users.sql        # rollback v2 schema
+- V2_1__Fix_topic_name.sql # A new version to migrate (v2.1)
+- U2_1__Fix_topic_name.sql # rollback v2.1 schema
+```
+
+It is recommended the user creates these two files on any new migration. The tool will not enforce that. We need specify this recomendation in the ksqlDB documents.
+
+### Command Syntax
+
+Finally, let's look at the rest of the command parameters that will exist to facilitate schema migrations.
+
+```shell script
+Usage:
+  ksql-migrations [options] commands
+  
+Commands
+  initialize   Initializes the schema version table
+
+  apply ( all | next | until <target> )
+  
+              Migrates a schema to new available schema versions (default: next)
+              
+              If 'all' is specified, then it migrates all newer versions available
+              If 'next' is specified, then it migrates only the next available version
+              If 'until <target>' is specified, then it migrates all available versions before <target>
+  
+  undo ( all | last | until <target> )
+  
+              Rollbacks a schema to the previous schema version (default: last)
+              
+              If 'all' is specified, then it rollbacks all previous versions
+              If 'last' is specified, then it rollbacks the previous version
+              If 'until <target>' is specified, then it rollbacks all previous versions after <target>
+  
+  info        Displays information about the current and available migrations
+  
+  baseline    Sets the current schema to the specified version
+  
+              This is useful when production environments already have a schema unversioned. This sets
+              the baseline from where to start applying migrations (i.e. kqsl-migrations baseline 1 'Current State')
+  
+  clean       Cleans the schema metadata objects                
+  
+  validate    Validate applied migrations against local files
+  
+              Compares local files checksum against the current metadata checksums to check for migrations files that have changed.
+              This tells the user that their schema might be not valid against their local files.
+  
+Options
+  -c, --config-file  Specifies a configuration file
+ 
+  -h, --help         Shows this help  
+    
+```
+
+### Configurations
+
+The tool will support a configuration file where all details for the server and tool migrations will be set. The following
+are the basic configurations that the tool should support.
+
+Example of `ksql-migrations.properties`
+```
+# Server URL and authentication
+ksql.server.url='http://localhost:8080'
+ksql.username='user1'
+ksql.password='pass1'
+
+# Migrations details
+ksql.migrations.stream.name='migration_events'
+ksql.migrations.table.name='schema_version'
+```
+
+Note: The command line options and other configurations will also be defined during the implementation.
+
+### Questions
```

Comment:
What would happen if multiple CI servers tried to migrate against the same cluster at the same time? Put differently, what sort of race conditions can happen?

Reply:
Locking the migrations stream and tables may help to avoid these race conditions. I need to think how to do that to also avoid locking an object during a failure in the migrations. Say the network connection fails, thus cannot unlock the stream and table.

I am thinking of adding another command to the tool to manually unlock or fix a failed migration. I'll add more info to the KLIP later.

Reply:
I considered using a second stream and table to store lock information, but it is getting complicated. I tried different ideas (which we can discuss offline), but I can't make it work. For instance, say we have a stream with schema `(id, locked, locked_by)` and the following rows:
```
id,  locked,  locked_by
* 1, true       u1
  1, true       u2
  1, true       u3
  1, false     u1
```
How can I create a table that shows me who has the lock? It should be `u1`. I was playing with `earliest_by_offset`, `latest_by_offset`. Also `latest_by_offset(id, 2)` to get the latest two to check in the application which one has the lock, but I forgot I could have more than 2 users trying to migrate. Without having transactions or a locking mechanism in ksqlDB, this is not reliable.

I think the only solution would be to support `LOCK` in ksqlDB, which requires session connections with the clients. This would let clients lock a table for the client session, thus avoiding other sessions read or write into a stream or table. @colinhicks We talked about sessions before, and I think we need this time. What do you think?

Reply:
Locking might get really complicated to add. We don't have to solve this right out of the gate, but we should know how it'll behave.